File: cpp/questions/coinChange.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        std::vector<int> dp(amount+1, amount+1);
        dp[0] = 0;
        for (const auto c: coins) {
            for (int i=c; i<amount+1; ++i) {
                dp[i] = std::min(dp[i], dp[i-c]+1);
            }
        }
        return dp[amount]==amount+1 ? -1 : dp[amount];
    }
};

File: python/questions/coinChange.py
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount+1] * (amount+1)
        dp[0] = 0
        for c in coins:
            for cs in range(c, amount+1):
                dp[cs] = min(dp[cs], dp[cs-c] + 1)
        return -1 if dp[amount]==amount+1 else dp[amount]
